clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,65 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+WhereClauseFact *Parser::ParseWhereClauseFact() {
+  // TODO: Handle bounds expression surrounded by parentheses, like:
+  // _Where ((((p : bounds(p, p + 1))))).
+  // Equality expressions surrounded by parentheses are already handled by
+  // invoking IgnoreValuePreservingCasts in ActOnEqualityOpFact.
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    SourceLocation IdLoc = ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Get the location of the start of bounds expr.
+    SourceLocation BoundsLoc = Tok.getLocation();
+
+    // Parse a bounds decl expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
```

Comment:
Here is a suggestion for how to improve the error messages generated for a where clause when a bounds declaration or bounds expression is missing.  This can be done separately.

The Sema phase in clang tracks the kind of scope that we are in, using a stack of scopes.  You could create a "where" clause scope.  Then in ParseExpr.cpp, you can check whether we are in a where clause to emit a different error message.  See clang\include\sema\clang\Scope.h.   See Parser::EnterScope and Parser::ExitScope in Parser.h.